Apps/Src/CommandLine.c

```diff
@@ -0,0 +1,357 @@
+	"determined contactor\n\r"
+	"	Contactors_Set array_c/array_p/motor_c on/off (non)blocking -\n\r"
+	"Sets the determined contactor\n\r"
+	"	Contactors_Enable  array_c/array_p/motor_c - Enables the determined\n\r"
+	"contactor\n\r"
+	"	Contactors_Disable  array_c/array_p/motor_c - Disables the determined\n\r"
+	"contactor\n\r"
+	"	Minion_Read_Input 'input' - Reads the current status of the input\n\r"
+	"	Minion_Write_Output `output` on/off - Sets the current state of the output\n\r"
+	"	Pedals_Read accel/brake - Reads the current status of the pedal\n\r"
+};
+
+static inline bool isWhiteSpace(char character){
+	switch (character) {
+		case 0x09:
+		case 0x0A:
+		case 0x0C:
+		case 0x0D:
+		case 0x20: return true;
+		default: return false;
+	}
+}
+
+static bool executeCommand(char *input) {
+	// The first word in the input should be a valid command
+	char *command = strtok_r(input, " ", &save);
+	// Iterate through all valid commands and check if the input matches
+	for (int i=0; cmdline_commands[i].name; i++) {
+		if (!strcmp(command, cmdline_commands[i].name)) {
+			return cmdline_commands[i].action(); // Execute the command
+		}
+	}
+	return false; // Didn't find a valid command!
+}
+
+// *********** Command line (shell) ************
+void Task_CommandLine(void* p_arg) { 
+	OS_ERR err;
+
+	// output welcome/help screen
+	printf(help);
+	
+	while(1){
+		printf("> ");
+		BSP_UART_Read(UART_2, input);
+		printf("\n\r");
+
+		if (!executeCommand(input)) { // If command failed, error
+			printf("Bad cmd. Please try again\n\r");
+		}
+	}
+
+    // Delay of 1 seconds
+    OSTimeDlyHMSM(0, 0, 1, 0, OS_OPT_TIME_HMSM_STRICT, &err);
+    if (err != OS_ERR_NONE){
+        assertOSError(OS_NONE_LOC, err);
+    }
+}
+
+// Function Implementations
+// ------------------------------------------------------------------------
+
+static inline bool cmd_help(void) {
+	printf(help);
+	return true;
+}
+
+/* This has not been tested and not sure if this implementation is correct (not sure
+if we want to just transmit string data from serial) */
+static bool cmd_CANbus_Send(void){
+	char *data = strtok_r(NULL, " ", &save);
+	CANDATA_t msg = {.ID=0x0582, .idx=0};		// this would change in the future (don't assume char as data)
+	for(int i = 0; i < 8 && i < strlen(data); i++){
+		msg.data[i] = data[i];
+	}
+
+	char *blockInput = strtok_r(NULL, " ", &save);
+	bool blocking;
+	if(strcmp(blockInput, "blocking") == 0){
+		blocking = CAN_BLOCKING;
+	}
+	else if(strcmp(blockInput, "nonblocking") == 0){
+		blocking = CAN_NON_BLOCKING;
+	}
+	else{
+		return false;
+	}
+
+	char *busInput = strtok_r(NULL, " ", &save);
+	CAN_t bus;
+	if(strcmp(busInput, "motor") == 0){
+		bus = CAN_1;
+	}
+	else if(strcmp(busInput, "car") == 0){
+		bus = CAN_3;
+	}
+	else{
+		return false;
+	}
+
+	if(CANbus_Send(msg, blocking, bus)){
+		printf("msg sent on %s (%s)\n\r", "can", blockInput);
+	}else{
+		printf("msg sent failed\n\r");
+	}
+	return true;
+}
+
+static bool cmd_CANbus_Read(void){
```

Comment:
I'm interested to see how this will interact with the readcarcan thread. If it causes some sort of weird behavior / data races, you might have to rethink this (if debug steals a message from ReadCarCAN or vice versa). An alternate approach could be for ReadCarCAN to have a getter available that returns the most recently read CAN message, just so that you can make sure that it sees all messages on the carCAN line. This is just a curiosity though, pure functionality looks fine.

Reply:
I think that might happen, but I will wait for testing to be done to confirm this. @Cam0Cow. I do like the alternative approach.

Reply:
Haven't had a chance to test this yet but I agree with Sid, we should cache the last CAN message received and read that.